File: win/Ratload.java
import jssc.SerialPort;
import jssc.SerialPortList;
import jssc.SerialPortException;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.FileReader;

import java.util.concurrent.Callable;
import java.util.concurrent.Executors;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.ThreadFactory;

public class Ratload {

   private static class HandshakeException extends Exception {
      private static final long serialVersionUID = 1L;
      public HandshakeException() {}

      @Override
      public String getMessage() {
         return "Handshake with Nexys board failed.";
      }
   }

   private static class InvalidProgRomException extends Exception {
      private static final long serialVersionUID = 1L;
      public InvalidProgRomException() {}

      @Override
      public String getMessage() {
         return "Invalid prog_rom file.";
      }
   }

   private static class SerialDataException extends Exception {
      private static final long serialVersionUID = 1L;
      public SerialDataException() {}

      @Override
      public String getMessage() {
         return "Received bad data from Nexys board.";
      }
   }

   private static class ByteRT implements Callable<Void> {

      private Thread mainThread;
      private SerialPort port;
      private byte b;
      
      public ByteRT(Thread mainThread, SerialPort port, byte b) {
         this.mainThread = mainThread;
         this.port = port;
         this.b = b;         
      }

      public Void call() throws
            SerialPortException,
            Ratload.SerialDataException {
         byte confirm;

         this.port.writeByte(this.b);
         confirm = port.readBytes(1)[0];
         if (confirm != this.b) {
            throw new Ratload.SerialDataException();
         }

         mainThread.interrupt();
         return null;
      }
   }

   private static final int PROG_ROM_LINES = 1024;
   private static final int PROG_ROM_SEGS = 5;
   private static final int INIT_HEIGHT = 64;
   private static final int INIT_WIDTH = 64;
   private static final int INITP_HEIGHT = 8;
   private static final int INITP_WIDTH = 64;
   private static final byte MAGIC_BYTE = 0X7E;
   private static final int TEST_LENGTH = 16;

   private static ExecutorService threadExec;

   public static void main(String[] args) {
      String[] parsed = parse_args(args);

      // Initialize the ExecutorService. This will manage the threads
      // that communicate with the serial port.
      threadExec = Executors.newSingleThreadExecutor(new ThreadFactory() {
         public Thread newThread(Runnable r) {
            Thread thread = new Thread(r);
            thread.setDaemon(true);
            return thread;
         }
      });
      
      // If we were using Java 7 we could use the fancy new syntax for
      // handling multiple exceptions the same way...
      // Instead we're stuck with this ridiculous try-catch block.
      try {
         if (parsed[0] == null) {
            run_serial_test(parsed[1]);
         } else {
            program_board(parsed[0], parsed[1]);
         }
      } catch (FileNotFoundException e) {
         System.err.println(e.getMessage());
      } catch (InvalidProgRomException e) {
         System.err.println(e.getMessage());
      } catch (IOException e) {
         System.err.println(e.getMessage());
      } catch (SerialPortException e) {
         System.err.println(e.getMessage());
      } catch (HandshakeException e) {
         System.err.println(e.getMessage());
      } catch (SerialDataException e) {
         System.err.println(e.getMessage());
      } catch (TimeoutException e) {
         System.err.println(e.getMessage());
      }

      threadExec.shutdownNow();
   }

   private static void run_serial_test(String serialDevice) throws
         SerialPortException,
         SerialDataException,
         TimeoutException,
         HandshakeException {
      SerialPort sp = new SerialPort(serialDevice);
      byte i, confirm;

      config_and_handshake(sp);
      
      for (i = 0; i < TEST_LENGTH; i++) {
         sp.writeByte((byte) int_to_char(i));
         confirm = sp.readBytes(1)[0];
         if (confirm != int_to_char(i)) {
            throw new Ratload.SerialDataException();
         }
      }

      System.out.println("PASS");
   }

   private static byte char_to_int(final char in) throws
         InvalidProgRomException {
      if (in >= '0' && in <= '9') {
         return (byte) (in - '0');
      } else if (in >= 'A' && in <= 'F') {
         return (byte) (in - 'A' + 10);
      } else {
         throw new InvalidProgRomException();
      }
   }
   
   private static void program_board(String file, String serialDevice) throws
         FileNotFoundException,
         InvalidProgRomException,
         IOException,
         SerialPortException,
         TimeoutException,
         SerialDataException,
         HandshakeException {
      byte[][] progRomArr = new byte[PROG_ROM_LINES][PROG_ROM_SEGS];
      char[][] progRomProper = new char[PROG_ROM_LINES][PROG_ROM_SEGS];
      byte b;
      int i, j;
      FileReader progRom = new FileReader(new File(file));
      SerialPort sp = new SerialPort(serialDevice);

      // Loop through the INIT prog_rom array.
      loop_to_array(progRom);
      for (i = 0; i < INIT_HEIGHT; i++) {
         for (j = 0; j < INIT_WIDTH; j++) {
            progRomArr[(i * 16) + ((63 - j) / 4)][j % 4 + 1] =
               char_to_int((char) progRom.read());
         } 

         // Consume the " at the end of the string.
         progRom.read();
         loop_to_array(progRom);
      }

      // Loop through the INITP prog_rom array.
      for (i = 0; i < INITP_HEIGHT; i++) {
         for (j = 0; j < INITP_WIDTH; j++) {
            b = char_to_int((char) progRom.read());
            progRomArr[(i * 128) + ((63 - j) * 2) + 1][0] = (byte) (((int) b & 0X0C) >> 2);
            progRomArr[(i * 128) + ((63 - j) * 2)][0] = (byte) ((int) b & 0x03);
         }

         progRom.read();
         loop_to_array(progRom);
      }

      progRom.close();

      // Convert the instructions *back* to ASCII.
      for (i = 0; i < PROG_ROM_LINES; i++) {
         for (j = 0; j < PROG_ROM_SEGS; j++) {
            progRomProper[i][j] = int_to_char(progRomArr[i][j]);
         }
      }

      config_and_handshake(sp);

      System.out.println("Connection opened: sending data...");

      // Send the instructions to the UART.
      for (i = 0; i < PROG_ROM_LINES; i++) {
         for (j = PROG_ROM_SEGS - 1; j > -1; j--) {
            bytert(sp, (byte) progRomProper[i][j]);
         }
         if (i % 103 == 0) {
            System.out.print((int) (i / 1024.0 * 100) + "% ");
         }
      }

      sp.closePort();
      System.out.printf("\nFinished.\n");
   }

   // The "handshaking procedure". Send a special byte to the
   // UART and wait for it to send it back. This will tell the
   // UART we are ready and confirm the UART is ready as well.
   private static void config_and_handshake(SerialPort port) throws
         Ratload.HandshakeException,
         SerialPortException,
         TimeoutException {
      port.openPort();
      port.setParams(57600, 8, 1, 1, false, false);

      try {
         bytert(port, MAGIC_BYTE);
      } catch (Ratload.SerialDataException e) {
         // Convert the SerialDataException to a HandshakeException
         // in this special case.
         throw new Ratload.HandshakeException();
      }
   }

   private static void bytert(SerialPort port, byte b) throws
         Ratload.SerialDataException,
         SerialPortException,
         TimeoutException {
      ByteRT rt = new ByteRT(Thread.currentThread(), port, b);
      Future<Void> res = null;
      Throwable unknown;

      try {
         res = threadExec.submit(rt);
         res.get(10, TimeUnit.MILLISECONDS);
      } catch (InterruptedException e) {
         // This means that the thread finished successfully.
         // We don't need to do anything here.
      } catch (ExecutionException e) {
         // The thread threw an exception; figure out what it was
         // and re-throw it.
         unknown = e.getCause();

         if (unknown instanceof Ratload.SerialDataException) {
            throw (Ratload.SerialDataException) unknown;
         } else {
            throw (SerialPortException) unknown;
         }
      } catch (TimeoutException e) {
         // Catch the TimeoutException, cancel the thread, and re-throw it.
         res.cancel(true);
         throw new TimeoutException("Timeout.");
      }
   }

   private static char int_to_char(final byte in) {
      return (char) (in + (in <= 9 ? '0' : ('A' - 10)));
   }

   // Loop to get to the beginning of the data.
   private static void loop_to_array(FileReader prog_rom) throws IOException {
      char a = ' ';

      while (a != '"') {
         a = (char) prog_rom.read();
      }
   }

   private static String[] parse_args(String[] args) {
      String[] parsed = new String[2], portNames;
      boolean fFound = false, dFound = false, lFound = false,
         hFound = false, tFound = false;
      int i;

      for (i = 0; i < args.length; i++) {
         if (i < args.length - 1 &&
             (args[i].equals("-f") || args[i].equals("--file"))) {
            fFound = true;
            parsed[0] = args[i + 1];
            i++;
         } else if (i < args.length - 1 &&
                    (args[i].equals("-d") || args[i].equals("--device"))) {
            dFound = true;
            parsed[1] = args[i + 1];
            i++;
         } else if (args[i].equals("-h") || args[i].equals("--help")) {
            hFound = true;
         } else if (args[i].equals("-l") || args[i].equals("--list")) {
            lFound = true;
         } else if (args[i].equals("-t") || args[i].equals("--test")) {
            tFound = true;
         }
      }

      if (!hFound && lFound) {
         portNames = SerialPortList.getPortNames();
         for (i = 0; i < portNames.length; i++) {
            System.out.println(portNames[i]);
         }
         System.exit(0);
      } else if (hFound || (!tFound && (!fFound || !dFound)) ||
                 (tFound && !dFound) || (tFound && fFound)) {
         System.out.println("Usage: ratload -d|--device <serial device> " +
                            "-f|--file <prog_rom file>");
         System.out.println("       (program Nexys2 board)");
         System.out.println("   or  ratload -d|--device <serial device> " +
                            "-t|--test");
         System.out.println("       (run serial connection test)");
         System.out.println("   or  ratload -l|--list");
         System.out.println("       (list available serial devices)");
         System.out.println("   or  ratload -h|--help");
         System.out.println("       (print this message)");
         System.exit(1);
      }

      if (tFound) {
         parsed[0] = null;
      }

      return parsed;
   }
}

File: nix/ratload.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>
#include <termios.h>
#include <fcntl.h>
#include <sys/types.h>

#include "ratload.h"

static void parse_args(char** parsed, const int argc, char* argv[]);
static int config_and_handshake(int* fd, const char* port);
static int program_board(const char* fileName, const char* portName);
static int run_serial_test(const char* portName);
static void loop_to_array(FILE * prog_rom);
static inline char int_to_char(const uint8_t in);
static int char_to_int(uint8_t* out, char in);

static const char* err2Str[NUM_ERRORS + 1] = {
   "",                                    // 0
   "File not found.",                     // 1
   "Invalid prog_rom file.",              // 2
   "Bad file read.",                      // 3
   "Bad serial device.",                  // 4
   "Configuring serial device failed.",   // 5
   "Handshake with Nexys board failed.",  // 6
   "Received bad data from Nexys board.", // 7
   "Timeout."                             // 8
};

int main(const int argc, char* argv[]) {
   int res;
   char* parsed[2] = {"", ""};

   parse_args(parsed, argc, argv);

   if (parsed[0] == NULL) {
      res = run_serial_test(parsed[1]);
   } else {
      res = program_board(parsed[0], parsed[1]);
   }

   if (res) {
      fprintf(stderr, "%s\n", err2Str[res]);
   }

   return 0;
}

static int run_serial_test(const char* portName) {
   uint8_t confirm, i;
   int sp, res;

   if ((res = config_and_handshake(&sp, portName))) {
      return res;
   }

   for (i = 0; i < TEST_LENGTH; i++) {
      confirm = int_to_char(i);
      if (write(sp, &confirm, 1) < 1) {
         close(sp);
         perror("write");
         return E_BAD_DEV;
      }

      if (read(sp, &confirm, 1) < 1) {
         close(sp);
         perror("read");
         return E_BAD_DEV;
      }

      if (confirm != int_to_char(i)) {
         return E_BAD_DATA;
      }
   }

   return 0;
}

static int program_board(const char* fileName, const char* portName) {
   uint8_t progRomArr[PROG_ROM_LINES][PROG_ROM_SEGS], b;
   char progRomProper[PROG_ROM_LINES][PROG_ROM_SEGS];
   FILE * progRom;
   char confirm;
   int i, j, fd, res;

   if ((progRom = fopen(fileName, "r")) == NULL) {
      return E_NO_FILE;
   }

   // Loop through the INIT prog_rom array.
   loop_to_array(progRom);
   for (i = 0; i < INIT_HEIGHT; i++) {
      for (j = 0; j < INIT_WIDTH; j++) {
         if ((res = char_to_int(&b, fgetc(progRom)))) {
            fclose(progRom);
            return res;
         }
         progRomArr[(i * 16) + ((63 - j) / 4)][j % 4 + 1] = b;
      }
      // The " at the end of the string.
      fgetc(progRom);
      loop_to_array(progRom);
   }

   // Loop through the INITP prog_rom array.
   for (i = 0; i < INITP_HEIGHT; i++) {
      for (j = 0; j < INITP_WIDTH; j++) {
         if ((res = char_to_int(&b, fgetc(progRom)))) {
            fclose(progRom);
            return res;
         }
         progRomArr[(i * 128) + ((63 - j) * 2) + 1][0] = (b & 0X0C) >> 2;
         progRomArr[(i * 128) + ((63 - j) * 2)][0] = b & 0x03;
      }
      fgetc(progRom);
      loop_to_array(progRom);
   }

   fclose(progRom);

   //convert the instructions BACK to ASCII
   for (i = 0; i < PROG_ROM_LINES; i++) {
      for (j = 0; j < PROG_ROM_SEGS; j++) {
         progRomProper[i][j] = int_to_char(progRomArr[i][j]);
      }
   }

   if ((res = config_and_handshake(&fd, portName))) {
      return res;
   }

   printf("Connection opened: sending data\n");

   //send the instructions to the UART
   for (i = 0; i < 1024; i++) {
      for (j = 4; j > -1; j--) {
         if (write(fd, &progRomProper[i][j], 1) < 1 ||
             read(fd, &confirm, 1) < 1) {
            perror("write");
            return E_BAD_DEV;
         }
      }
      if (!(i % 103)) {
         printf("%d%% ", (int) (i / 1024.0 * 100));
         fflush(stdout);
      }
   }

   close(fd);
   printf("\nFinished.\n");

   return 0;
}

static int config_and_handshake(int* sp, const char* portName) {
   struct termios options;
   char confirm = MAGIC_BYTE;

   if ((*sp = open(portName, O_RDWR)) == -1) {
      perror("open");
      return E_BAD_DEV;
   }

   // Configuration of the serial port.
   if (tcgetattr(*sp, &options) == -1) {
      perror("tcgetattr");
      return E_CONF_FAIL;
   }

   cfsetispeed(&options, B57600);
   cfsetospeed(&options, B57600);
   //options.c_cflag = CLOCAL | CREAD | PARENB | PARODD | CS8;
   options.c_cflag = CLOCAL | CREAD | PARENB | PARODD | CS8;
   options.c_iflag = 0;
   options.c_oflag = 0;
   options.c_lflag = ICANON;

   if (tcsetattr(*sp, TCSANOW, &options) == -1) {
      close(*sp);
      perror("tcsetattr");
      return E_CONF_FAIL;
   }

   fprintf(stderr, "HERE1 sp is %d\n", *sp);

   if (write(*sp, &confirm, 1) < 1) {
      close(*sp);
      perror("write");
      return E_BAD_DEV;
   }
   fprintf(stderr, "HERE2 sp is %d\n", *sp);

   confirm = 0;

   if (read(*sp, &confirm, 1) < 1) {
      close(*sp);
      perror("read");
      return E_BAD_DEV;
   }

   fprintf(stderr, "HERE3\n");

   if (confirm != MAGIC_BYTE) {
      close(*sp);
      return E_HANDSHAKE;
   }

   return 0;
}

static void parse_args(char** parsed, const int argc, char* argv[]) {
   bool fFound = false, dFound = false, lFound = false,
      hFound = false, tFound = false;
   int i;

   for (i = 0; i < argc; i++) {
      if (i < argc - 1 &&
          (!strcmp(argv[i], "-f") || !strcmp(argv[i], "--file"))) {
         fFound = true;
         parsed[0] = argv[i + 1];
         i++;
      } else if (i < argc - 1 &&
                 (!strcmp(argv[i], "-d") || !strcmp(argv[i], "--device"))) {
         dFound = true;
         parsed[1] = argv[i + 1];
         i++;
      } else if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
         hFound = true;
      } else if (!strcmp(argv[i], "-l") || !strcmp(argv[i], "--list")) {
         lFound = true;
      } else if (!strcmp(argv[i], "-t") || !strcmp(argv[i], "--test")) {
         tFound = true;
      }
   }

   if (!hFound && lFound) {
      printf("Serial port listing not supported on this platform.\n");
      exit(EXIT_SUCCESS);
   } else if (hFound || (!tFound && (!fFound || !dFound)) ||
              (tFound && !dFound) || (tFound && fFound)) {
      printf("Usage: ratload -d|--device <serial device> "
             "-f|--file <prog_rom file>\n"
             "       (program Nexys2 board)\n"
             "   or  ratload -d|--device <serial device> -t|--test\n"
             "       (run serial connection test)\n"
             "   or  ratload -l|--list\n"
             "       (list available serial devices)\n"
             "   or  ratload -h|--help\n"
             "       (print this message)\n");
      exit(EXIT_FAILURE);
   }

   if (tFound) {
      parsed[0] = NULL;
   }
}

//loop to get to the beginning of the data
static void loop_to_array(FILE * progRom) {
   char a = ' ';

   while (a != '"') {
      a = fgetc(progRom);
   }
}

static inline char int_to_char(const uint8_t in) {
   return in + (in <= 9 ? '0' : ('A' - 10));
}

static int char_to_int(uint8_t* out, char in) {
   int err = 0;

   if (in >= '0' && in <= '9') {
      *out = in - '0';
   } else if (in >= 'A' && in <= 'F') {
      *out = in - 'A' + 10;
   } else {
      err = E_BAD_FILE;
   }

   return err;
}
